it/src/test/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManagerTest.java

```diff
@@ -0,0 +1,236 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.it.pubsub;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.cloud.pubsub.v1.SubscriptionAdminClient;
+import com.google.cloud.pubsub.v1.TopicAdminClient;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PushConfig;
+import com.google.pubsub.v1.Subscription;
+import com.google.pubsub.v1.SubscriptionName;
+import com.google.pubsub.v1.Topic;
+import com.google.pubsub.v1.TopicName;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+/** Unit tests for {@link DefaultPubsubResourceManager}. */
+@RunWith(JUnit4.class)
+public final class DefaultPubsubResourceManagerTest {
+  @Rule public final MockitoRule mockito = MockitoJUnit.rule();
+
+  private static final String TEST_ID = "test-id-";
+  private static final String PROJECT_ID = "test-project";
+  private static final String TOPIC_NAME = "test-topic-name";
+  private static final String SUBSCRIPTION_NAME = "test-topic-name-sub0";
+  private static final TopicName TOPIC_REFERENCE = TopicName.of("test-project", "test-topic");
+  private static final String VALID_MESSAGE_ID = "abcdef";
+
+  @Mock private TopicAdminClient topicAdminClient;
+  @Mock private SubscriptionAdminClient subscriptionAdminClient;
+  @Mock private Topic topic;
+  @Mock private Subscription subscription;
+  @Mock private Publisher publisher;
+
+  private DefaultPubsubResourceManager testManager;
+
+  @Captor private ArgumentCaptor<TopicName> topicNameCaptor;
+  @Captor private ArgumentCaptor<SubscriptionName> subscriptionNameCaptor;
+  @Captor private ArgumentCaptor<PubsubMessage> pubsubMessageCaptor;
+
+  @Before
+  public void setUp() throws IOException {
+    testManager =
+        spy(
```

Comment:
Is there a reason for making this a spy?

Reply:
Yes, to inject our mocked Publisher on `getPublisher(topic)`:
```
    doReturn(publisher).when(testManager).getPublisher(any());
```

Reply:
We generally do dependency injection through the constructor so that the class under test is a real object. If there's a bug in `getPublisher`, this would hide that. Is there something about the publisher that prevents us from making it a field?

Reply:
Publishers are created for a specific topic/subscription, which are not known ahead of time.

Reply:
I think then it makes sense to have a package-private `PublisherFactory` class that handles this and inject that. I think it would be better to inject that mock than use a spy.